contracts/account/paymaster/PaymasterERC20.sol

```diff
@@ -0,0 +1,115 @@
+    function _validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 maxCost
+    ) internal virtual override returns (bytes memory context, uint256 validationData) {
+        (IERC20 token, uint48 validAfter, uint48 validUntil, uint256 tokenPrice, address guarantor) = _fetchDetails(
+            userOp,
+            userOpHash
+        );
+
+        uint256 prefundAmount = (maxCost +
+            (guarantor == address(0)).ternary(POST_OP_COST, POST_OP_COST_WITH_GUARANTOR) *
+            userOp.maxFeePerGas()).mulDiv(tokenPrice, _tokenPriceDenominator());
+
+        return (
+            abi.encodePacked(userOpHash, token, prefundAmount, tokenPrice, userOp.sender, guarantor),
+            token
+                .trySafeTransferFrom(guarantor == address(0) ? userOp.sender : guarantor, address(this), prefundAmount)
+                .packValidationData(validAfter, validUntil)
+        );
+    }
+
+    /// @inheritdoc PaymasterCore
+    function _postOp(
+        PostOpMode /* mode */,
+        bytes calldata context,
+        uint256 actualGasCost,
+        uint256 actualUserOpFeePerGas
+    ) internal virtual override {
+        bytes32 userOpHash = bytes32(context[0x00:0x20]);
+        IERC20 token = IERC20(address(bytes20(context[0x20:0x34])));
+        uint256 prefundAmount = uint256(bytes32(context[0x34:0x54]));
+        uint256 tokenPrice = uint256(bytes32(context[0x54:0x74]));
+        address user = address(bytes20(context[0x74:0x88]));
+        address guarantor = address(bytes20(context[0x88:0x9C]));
+
+        uint256 actualAmount = (actualGasCost +
+            (guarantor == address(0)).ternary(POST_OP_COST, POST_OP_COST_WITH_GUARANTOR) *
+            actualUserOpFeePerGas).mulDiv(tokenPrice, _tokenPriceDenominator());
+
+        if (guarantor == address(0)) {
+            token.safeTransfer(user, prefundAmount - actualAmount);
+            emit UserOperationSponsored(userOpHash, user, address(0), actualAmount, tokenPrice, false);
+        } else if (token.trySafeTransferFrom(user, address(this), actualAmount)) {
+            token.safeTransfer(guarantor, prefundAmount);
+            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, false);
+        } else {
+            token.safeTransfer(guarantor, prefundAmount - actualAmount);
+            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, true);
+        }
+    }
+
+    /**
+     * @dev Internal function that returns the repayment details for a given user operation
+     *
+     * This may be implemented in any number of ways, including
+     * * Hardcoding values (only one token supported)
+     * * Getting the price from an onchain oracle
+     * * Getting the (signed) values through the userOp's paymasterData
+     *
+     * The paymaster can also decide to not support guarantors, and always return address(0) for that part.
```

Comment:
Following my previous comment, I think this makes it clearer

```suggestion
     * The paymaster can also decide to not support guarantors, and always return address(0) for that part.
     *
     * NOTE: If a guarantor is supported, make sure that it can't be used arbitrarily to pay operations.
     * Concretely, if the guarantor is extracted from the `userOp`, then it'll be possible to use any
     * address who's approved this contract
```

Reply:
How would you fell about: 

```suggestion
     * The paymaster can also decide to not support guarantors, and always return address(0) for that part.
     *
     * NOTE: If a guarantor is supported, make sure that it can't be used arbitrarily to pay operations.
     * Concretely, if the guarantor is extracted from the `userOp`, make sure that it provided explicit consent to 
     * support that user operation, for example by verifying a signature.
```